web/src/upload.ts

```diff
@@ -254,6 +254,59 @@ export function rewire_upload_files(value: typeof upload_files): void {
     upload_files = value;
 }
 
+type PreviousUpload = {
+    size: number | null;
+    metadata: Record<string, string>;
+    creationTime: string;
+    urlStorageKey: string;
+    uploadUrl: string | null;
+    parallelUploadUrls: string[] | null;
+};
+
+// Parts of it are inspired from WebStorageUrlStorage at
+// https://github.com/tus/tus-js-client/blob/ca63ba254ea8766438b9d422f6f94284911f1fa5/lib/browser/urlStorage.js#L27
+// While there are no async actions happening in any of the methods in
+// this class, UrlStorage interface for tus-js-client requires a Promise
+// to be returned for each of these methods.
+class InMemoryUrlStorage {
+    urlStorage: Map<string, PreviousUpload>;
+
+    constructor() {
+        this.urlStorage = new Map();
+    }
+
+    async findAllUploads(): Promise<PreviousUpload[]> {
+        return await Promise.resolve([...this.urlStorage.values()]);
+    }
+
+    async findUploadsByFingerprint(fingerprint: string): Promise<PreviousUpload[]> {
+        const results = [];
+
+        for (const [key, value] of this.urlStorage) {
+            if (!key.startsWith(`${fingerprint}::`)) {
+                continue;
+            }
+            results.push(value);
+        }
+
+        return await Promise.resolve(results);
+    }
+
+    async removeUpload(urlStorageKey: string): Promise<void> {
+        this.urlStorage.delete(urlStorageKey);
+        await Promise.resolve();
+    }
+
+    async addUpload(fingerprint: string, upload: PreviousUpload): Promise<string> {
+        const id = Math.round(Math.random() * 1e12);
+        const key = `${fingerprint}::${id}`;
+
+        upload.urlStorageKey = key;
+        this.urlStorage.set(key, upload);
```

Comment:
the `key` and `urlStorageKey` are duplicated information. I looked into just using an array as the urlStorage. In that case, I would have to check whether a urlStorageKey existed or not, and replace that entry with the new one if it did. Which would have been a fine approach, but did not look that elegant, so kept on using maps. Let me know if we want to use a list as storage instead.